### PR TITLE
SCELoss(SCELossGrad) support half(float) input float(half) output

### PR DESCRIPTION
**Pull request by pengwa**
### Description

A follow up change for https://github.com/microsoft/onnxruntime/pull/13616. 

SoftmaxCrossEntropyLossInternal/SoftmaxCrossEntropyLossInternalGrad support different type for input and output.

Add SCELoss(SCELossGrad) support half(float) input float(half) output

#### Add tests for variant input and output types. To add such tests, have to refactor existing testing code for sce loss and scelossinternal gradient. 

Originally, 

FP32 input and output, the CPU kernels, runs with CPU kernels the baseline, CUDA/RCOM then runs with same data, user CompareTester to compare with CPU run results. 

FP16 input and output, the CPU kernels (did not have half kernels), runs with Cast_to_float->CPU kernel->cast_to_half as the baseline, CUDA/RCOM then runs with same data but using Half implementation, user CompareTester to compare with CPU run results. 

Now, we want the support run different input and output types. The proposed change here is, to run CPU kernels always with float input and output as baseline (because CPU only have float type kernels impl), this step is the very first thing for every test. 

Then, we run CUDA/RCOM kernels using half_input_half_output, float_input_float_output, half_input_float_output, float_input_half_output if there is corresponding kernel registered. 

Afterwards, compare the CUDA/ROCM run results with CPU float baselines. 

Be noted, there is one thing that deserved a special note: CompareOpTester's result compare can be loose than OpTester's. 
Roughly speaking: the former tolerant diff <= atol + rtol*expected_value, while the later one telerant diff < atol && diff < rtol*expected_value. When the expected value is super small in many cases of our tests cases, the former one can pass but the later one fails. So the refactoring also move the check outside of OpTester, explicitly check the values using the way CompareOPTester did (to align the previous behaviour).

### CODE REVIEW

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 1]
```diff
@@ -149,39 +149,97 @@ TensorProto ToDimensionOneTensor(T value) {
...
+  static std::unordered_map<::ONNX_NAMESPACE::TensorProto_DataType, int> bytes_for_elem_type = {
+      {ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT16, 2},
+      {ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_BFLOAT16, 2},
+      {ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT, 4},
+      {ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_DOUBLE, 8},
+  };
+  bool use_input_elem_type_for_compute = false;
+  // bytes_for_elem_type[static_cast<::ONNX_NAMESPACE::TensorProto_DataType>(input_elem_type)] >=
+  // bytes_for_elem_type[static_cast<::ONNX_NAMESPACE::TensorProto_DataType>(output_elem_type)];
+  return std::make_pair(use_input_elem_type_for_compute,
```
Comment (askhade) [inline orttraining/orttraining/core/graph/training_op_defs.cc:163]: use_input_elem_type_for_compute is always false... does not look intentional. Can you update the code here. Thanks!

Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:163]: Sorry, my bad, let me do a clean up and remove those used for debugging.

Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc]: Have refine it better. And also remove previous workaround fix. Thanks a lot!

**Review by askhade — APPROVED** [review]
LGTM

**Comment by askhade** [comment]
Is there a perf improvement going from the workaround to this solution?

**Comment by pengwa** [comment]
> Is there a perf improvement going from the workaround to this solution?

I think there should be, let me do a quick profiling to see whether it is obvious.

**Comment by pengwa** [comment]
> > Is there a perf improvement going from the workaround to this solution?
> 
> I think there should be, let me do a quick profiling to see whether it is obvious.

Here is a perf benchmarking: python -m torch.distributed.launch --nproc_per_node=4 examples/onnxruntime/training/language-modeling/run_mlm.py  --model_name_or_path roberta-base --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1  --num_train_epochs 10 --per_device_train_batch_size 8 --per_device_eval_batch_size 8 --do_train --do_eval --overwrite_output_dir --output_dir ./outputs/ --seed 1137 --fp16 --report_to none --optim adamw_ort_fused --max_steps 400 --logging_steps 20

image


The throughput increased a little bit: 100.052/99.454, 0.6% gain.
